Coloring diagnostics does not always work with clang

### How to reproduce ###
<!-- Steps to reproduce the behavior: -->
1. Create file `foo.cc` with below content:
```
int main() {
    return "";
}
```
2. Run: `ccache clang++ -Xclang -fcolor-diagnostics -c foo.cc 2>&1 | cat`

### Actual behavior ###
<!-- What happened? -->
ccache strips colors from diagnostics, here is an interesting excerpt from running it under strace which shows that clang did produce colored diagnostics, but ccache removed the colors:
```
openat(AT_FDCWD, "/run/user/1000/ccache-tmp/tmp.stderr.yQUHE0", O_RDONLY) = 4
read(4, "\33[1mfoo.cc:2:8: \33[0m\33[0;1;31merr"..., 1024) = 187
close(4)                                = 0
write(2, "foo.cc:2:8: error: cannot initia"..., 145foo.cc:2:8: error: cannot initialize return object of type 'int' with an lvalue of type 'const char [1]'
return "";
       ^~
1 error generated.
) = 145
```

### Expected behavior ###
<!-- What did you expect to happen? -->
ccache should print colored diagnostics

### Environment ###
<!-- Which ccache version did you use? -->

ccache version 4.2

<!-- If applicable, add more info such as compiler, OS and file system below: -->
Linux archlinux 5.10.16-arch1-1 SMP PREEMPT Sat, 13 Feb 2021 20:50:18 +0000 x86_64 GNU/Linux
clang version 11.0.1 (also reproduced with 10.0.0)

### Remarks ###
If I remove `-Xclang` or `-c` it works i.e. both `ccache clang++ -fcolor-diagnostics -c foo.cc 2>&1 | cat` and `ccache clang++ -Xclang -fcolor-diagnostics foo.cc 2>&1 | cat` give colored diagnostics.

--- a/src/argprocessing.cpp
+++ b/src/argprocessing.cpp
@@ -700,6 +700,14 @@ process_arg(Context& ctx,
     return Statistic::unsupported_compiler_option;
   }
 
+  // In the "-Xclang -fcolor-diagnostics" form, -Xclang is skipped and the
+  // -fcolor-diagnostics argument which is passed to cc1 is handled below.
+  if (args[i] == "-Xclang" && i < args.size() - 1
+      && args[i + 1] == "-fcolor-diagnostics") {
+    state.compiler_only_args_no_hash.push_back(args[i]);
+    ++i;
+  }
+
   if (args[i] == "-fcolor-diagnostics" || args[i] == "-fdiagnostics-color"
       || args[i] == "-fdiagnostics-color=always") {
     state.color_diagnostics = ColorDiagnostics::always;

--- a/test/suites/color_diagnostics.bash
+++ b/test/suites/color_diagnostics.bash
@@ -137,6 +137,15 @@ color_diagnostics_test() {
         fi
     fi
 
+    if $COMPILER_TYPE_CLANG; then
+        # ---------------------------------------------------------------------
+        TEST "-fcolor-diagnostics works when passed to cc1 with -Xclang"
+
+        color_diagnostics_generate_code test1.c
+        $CCACHE_COMPILE -Xclang -fcolor-diagnostics -Wreturn-type -c -o test1.o test1.c 2>test1.stderr
+        color_diagnostics_expect_color test1.stderr
+    fi
+
     while read -r case; do
         # ---------------------------------------------------------------------
         TEST "Cache object shared across ${case} (run_second_cpp=$run_second_cpp)"

--- a/unittest/test_argprocessing.cpp
+++ b/unittest/test_argprocessing.cpp
@@ -570,6 +570,8 @@ TEST_CASE("-Xclang")
     "-Xclang -fno-pch-timestamp"
     " -Xclang unsupported";
 
+  const std::string color_diag = "-Xclang -fcolor-diagnostics";
+
   const std::string extra_args =
     "-Xclang -emit-pch"
     " -Xclang -emit-pth";
@@ -580,8 +582,9 @@ TEST_CASE("-Xclang")
     " -Xclang -include-pth pth_path1"
     " -Xclang -include-pth -Xclang pth_path2";
 
-  ctx.orig_args = Args::from_string("gcc -c foo.c " + common_args + " "
-                                    + extra_args + " " + pch_pth_variants);
+  ctx.orig_args =
+    Args::from_string("gcc -c foo.c " + common_args + " " + color_diag + " "
+                      + extra_args + " " + pch_pth_variants);
   Util::write_file("foo.c", "");
 
   const ProcessArgsResult result = process_args(ctx);
@@ -589,8 +592,8 @@ TEST_CASE("-Xclang")
         == "gcc " + common_args + " " + pch_pth_variants);
   CHECK(result.extra_args_to_hash.to_string() == extra_args);
   CHECK(result.compiler_args.to_string()
-        == "gcc " + common_args + " " + extra_args + " " + pch_pth_variants
-             + " -c");
+        == "gcc " + common_args + " " + color_diag + " " + extra_args + " "
+             + pch_pth_variants + " -c");
 }
 
 TEST_CASE("-x")

